lib/Onnxifi/Base.cpp

```diff
@@ -48,22 +48,84 @@ onnxStatus Graph::initGraph(const void *onnxModel, size_t onnxModelSize,
   // TODO: support multiple functions here.
   function_ = backendPtr_->getEE().getModule().createFunction("inference");
 
-  std::unique_ptr<ModelLoader> loader = ModelLoader::parse(
-      onnxModel, onnxModelSize, weightCount, weightDescriptors, *function_);
+  // Keep loader to have an access to the input and output ONNX vars.
+  // TODO: The loader itself is not strictly required, remove that
+  // and just keep mapping.
+  loader_ = ModelLoader::parse(onnxModel, onnxModelSize, weightCount,
+                               weightDescriptors, *function_);
   // TODO: make better error reporting.
-  if (!loader) {
+  if (!loader_) {
     return ONNXIFI_STATUS_INTERNAL_ERROR;
   }
 
+  // Emit IR for the graph and compile it.
+  backendPtr_->getEE().compile(CompilationMode::Infer, function_);
+
   return ONNXIFI_STATUS_SUCCESS;
 }
 
-onnxStatus Graph::run() { return ONNXIFI_STATUS_SUCCESS; }
+onnxStatus Graph::run() {
+  // Copy tensors from the input addresses to the Glow tensors.
+  std::vector<Tensor *> tensors;
+  std::vector<glow::Variable *> vars;
+  for (auto inputVar : inputs_) {
+    auto *var = inputVar.first;
+    std::cout << "Run: input variable name: " << var->getName().str() << std::endl;
+    
+    auto *type = var->getType();
+    void *inputBuffer = reinterpret_cast<void *>(inputVar.second);
+    tensors.push_back(
+        new Tensor(inputBuffer, type));
+    vars.push_back(var);
+  }
+
+  // Run inference.
+  backendPtr_->getEE().run(vars, tensors);
+
+  // Copy outputs to the addresses specified in the outputs_.
+  for (auto outputVar : outputs_) {
+    void *outputAddress = reinterpret_cast<void *>(outputVar.second);
+    Tensor &res = outputVar.first->getVariable()->getPayload();
+
+    memcpy(outputAddress, res.getUnsafePtr(), res.size() * res.getType().getElementSize());
+  }
+
+  return ONNXIFI_STATUS_SUCCESS;
+}
 
 onnxStatus Graph::setIO(uint32_t inputsCount,
                         const onnxTensorDescriptorV1 *inputDescriptors,
                         uint32_t outputsCount,
                         const onnxTensorDescriptorV1 *outputDescriptors) {
+  // Process inputs.
```

Comment:
You should check tags before you access any members in `inputDescriptors` or `outputDescriptors` and return `ONNXIFI_STATUS_UNSUPPORTED_TAG` if tag does not match `ONNXIFI_TAG_TENSOR_DESCRIPTOR_V1`. Because if `tag` is not `ONNXIFI_TAG_TENSOR_DESCRIPTOR_V1`, `inputDescriptors[i]` is not a `onnxTensorDescriptorV1`, and even reading any member may SEGFAULT.